docs/rules/unicode-property.md

```diff
@@ -0,0 +1,241 @@
+---
+pageClass: "rule-details"
+sidebarDepth: 0
+title: "regexp/unicode-property"
+description: "enforce consistent naming of unicode properties"
+---
+# regexp/unicode-property
+
+🔧 This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).
+
+<!-- end auto-generated rule header -->
+
+> enforce consistent naming of unicode properties
+
+## :book: Rule Details
+
+This rule helps to enforce consistent style and naming of unicode properties.
+
+There are many ways a single Unicode property can be expressed. E.g. `\p{L}`, `\p{Letter}`, `\p{gc=L}`, `\p{gc=Letter}`, `\p{General_Category=L}`, and `\p{General_Category=Letter}` are all equivalent. This rule can be configured in a variety of ways to control exactly which ones of those variants are allowed. The default configuration is intended to be a good starting point for most users.
+
+<eslint-code-block fix>
+
+```js
+/* eslint regexp/unicode-property: "error" */
+
+/* ✓ GOOD */
+var re = /\p{L}/u;
+var re = /\p{Letter}/u;
+var re = /\p{Script=Greek}/u;
+var re = /\p{scx=Greek}/u;
+var re = /\p{Hex}/u;
+var re = /\p{Hex_Digit}/u;
+
+/* ✗ BAD */
+var re = /\p{gc=L}/u;
+var re = /\p{General_Category=Letter}/u;
+var re = /\p{Script=Grek}/u;
+```
+
+</eslint-code-block>
+
+## :wrench: Options
+
+```json
+{
+  "regexp/unicode-property": ["error", {
+    "generalCategory": "never",
+    "key": "ignore",
+    "property": {
+      "binary": "ignore",
+      "generalCategory": "ignore",
+      "script": "long",
+    }
+  }]
+}
+```
+
+### `generalCategory: "never" | "always" | "ignore"`
+
+Values from the `General_Category` property can be expressed in two ways: either without or with the `gc=` (or `General_Category=`) prefix. E.g. `\p{Letter}` or `\p{gc=Letter}`.
+
+This option controls whether the `gc=` prefix is required or forbidden.
+
+- `"never"` (default): The `gc=` (or `General_Category=`) prefix is forbidden.
+  <eslint-code-block fix>
+
+  ```js
+  /* eslint regexp/unicode-property: ["error", { generalCategory: "never" }] */
+
+  var re = /\p{Letter}/u;
+  var re = /\p{gc=Letter}/u;
+  var re = /\p{General_Category=Letter}/u;
+  ```
+
+  </eslint-code-block>
+
+- `"always"`: The `gc=` (or `General_Category=`) prefix is required.
+  <eslint-code-block fix>
+
+  ```js
+  /* eslint regexp/unicode-property: ["error", { generalCategory: "always" }] */
+
+  var re = /\p{Letter}/u;
+  var re = /\p{gc=Letter}/u;
+  var re = /\p{General_Category=Letter}/u;
+  ```
+
+  </eslint-code-block>
+
+- `"ignore"`: Both with and without prefix is allowed.
+  <eslint-code-block fix>
+
+  ```js
+  /* eslint regexp/unicode-property: ["error", { generalCategory: "ignore" }] */
+
+  var re = /\p{Letter}/u;
+  var re = /\p{gc=Letter}/u;
+  var re = /\p{General_Category=Letter}/u;
+  ```
+
+  </eslint-code-block>
+
+### `key: "short" | "long" | "ignore"`
```

Comment:
Should we support "medium"? https://unicode-org.github.io/icu/userguide/strings/unicodeset.html#property-values

Reply:
We already do with `generalCategory: "always", key: "short"`. This will enforce `\p{gc=X}`. So it's already possible to configure this rule to act like the "medium" from the page you linked.

I also don't think that a "medium" option really makes sense, because it only applies to the `General_Category` property as it is the only property (in JS) that can have 3 forms (`\p{X}`, `\p{gc=X}`, `\p{General_Category=X}`).